pkgs/top-level/all-packages.nix

```diff
@@ -19100,6 +19100,9 @@ with pkgs;
 
   inotify-tools = callPackage ../development/tools/misc/inotify-tools { };
 
+  inherit (callPackages ../applications/misc/inochi2d { })
```

Comment:
You may want `callPackage` (no `s`) here to allow individually overriding these packages.

Reply:
I believe `callPackages` is the correct thing to use, however I'm using `callPackage` internally (which I shouldn't), so it only allows me to override the things at the top of `default.nix`, which is not a lot. I'll see what I can do.

Reply:
I inverted the way the packages are placed: now the `mkGeneric` function resides in `default.nix`, while the package-specific stuff have been moved to their respective `.nix` files. Those files are only `import`-ed and not `callPackage`-ed now, so I need to manually pass down the dependencies with `inherit` from `default.nix`.

Reply:
> I believe `callPackages` is the correct thing to use, however I'm using `callPackage` internally (which I shouldn't), so it only allows me to override the things at the top of `default.nix`, which is not a lot. I'll see what I can do.

That's how callPackages is supposed to be used AFAICT.

I'm not sure I like the current iteration better tbh. Especially the manual passing of args is weird; that's what we have callPackage for. Take a look at emacs or ffmpeg if you want to see how that works.

Reply:
I moved everything into `default.nix` for now.

I *could* do something similar to `ffmpeg`'s method (passing the builder arguments through `callPackage`'s second argument), though I wouldn't be able to do `args //` easily as `args` would get polluted by all other dependencies, meaning the inputs would have to be `inherit`-ed instead of `//`-ed.

Reply:
Sorry, what I meant is how you had it before (when I made the comment) but simply s/callPackages/callPackage/. Keep the individual packages in separate files and callPackage them again internally rather than importing.

Reply:
That wouldn't make it any better for overriding: you can only mark one attribute set as overridable, and they don't propagate.
The way it is now, we can override anything.
If I move back to how it was previously, we'll only be able to override the non-generic inputs, which is not useful.

Though, I'll gladly do it your way if you can show me how it can be done.

Reply:
I might have come up with something, which I'll commit without squashing, to see what your opinion is about it.

Reply:
Now that I implemented it, it doesn't feel that bad, because the non-generic inputs only have helper stuff.

Reply:
Here's the diff from before moving everything around: https://github.com/NixOS/nixpkgs/compare/45815909412fccf1ccc56f4b0946072935f0660e..9a005d97f3597e789a1c748304036a773f08df61

Reply:
The way it'd work is that you have a default.nix which in turn `callPackage`'s the individual packages and the builder.

This default.nix needs some generic arguments such as `callPackage` itself and lib perhaps. It'd look a little like this:

```Nix
{
  lib,
  callPackage,
}:

{
  inochi-creator = callPackage ./creator.nix { };
  inochi-session = callPackage ./creator.nix { };
  mkGeneric = callPackage ./generic.nix { };
}
```

You'd pass `mkGeneric` to the other packages by either making it a `rec` attrset or I believe you could also simply create a new scope which I think is what Emacs does.

What I hadn't considered though is that the `mkGeneric` abstraction basically contains all the deps worth overriding.

I have a new idea though:

```Nix
{
  lib,
  buildDubPackage,
  fetchFromGitHub,
  writeShellScriptBin,

  ...
}:

let
  mkGeneric = ...;
in

{
  inochi-creator = mkGeneric ./creator.nix;
  inochi-session = mkGeneric ./creator.nix;
  inherit mkGeneric;
}
```

I'd still prefer if mkGeneric was somehow outside of this file but that'd be a bit tricky; requiring passing through the args.

Reply:
I think we're well past the point of ready though; so I'd be in favour of reverting to callPackages and calling it a day :)

Reply:
The original `callPackages` way didn't allow overriding properly. The one-file version with `callPackages` did allow overriding everything, but was a bit long. The current version allows overriding the most important stuff, while putting the generic stuff into a different file. This is how ffmpeg does it I believe.
I'll put some more thought into it and try to settle for something.

Reply:
I went with the current implementation, as that seems to be the most used method by other packages that have a separate `generic.nix`.

Reply:
I have 1 other method, which could work, used by e.g the `ed` package:
`callPackage` doesn't require a file as an input, it can take in a function directly, so we can first do `import` then `callPackage`.

So we could have `generic.nix` take be a nested function, the outer function takes in the per-package args, the inner function args get populated by `callPackage`

```nix
{
  inochi-session = callPackage (import ./generic.nix {
    pname = "inochi-session";
    ...
  });
}
```

Reply:
Oh, yeah that's a neat pattern. I don't see a reason why we wouldn't be able to put the creator and session into separate files though:

```Nix
let
  mkGeneric = builderArgs: callPackage ./generic.nix { inherit builderArgs; };
in
{
  inochi-creator = import ./creator.nix;
  inochi-session = import ./session.nix;
}
```

Reply:
Ah, wait the files need some generic deps such as lib and stdenv. Those would have to be passed explicitly (wouldn't be the end of the world) or I think you could just callPackage rather than import. The second callPackage should override the override functions I think?

Reply:
Tried moving the per-package args into their separate files, unfortunately, populating their dependencies with `callPackage` also adds the `.override` property, which would get `//`'d into the generic builder's args, which gives an error.